File: app/src/main/java/dev/ragnarok/fenrir/domain/impl/FeedbackInteractor.java
package dev.ragnarok.fenrir.domain.impl;

import static dev.ragnarok.fenrir.domain.mappers.Entity2Model.fillCommentOwnerIds;
import static dev.ragnarok.fenrir.util.Utils.isEmpty;

import java.util.ArrayList;
import java.util.List;

import dev.ragnarok.fenrir.api.interfaces.INetworker;
import dev.ragnarok.fenrir.api.model.feedback.VkApiBaseFeedback;
import dev.ragnarok.fenrir.db.interfaces.IStorages;
import dev.ragnarok.fenrir.db.model.IdPairEntity;
import dev.ragnarok.fenrir.db.model.entity.OwnerEntities;
import dev.ragnarok.fenrir.db.model.entity.feedback.CopyEntity;
import dev.ragnarok.fenrir.db.model.entity.feedback.FeedbackEntity;
import dev.ragnarok.fenrir.db.model.entity.feedback.LikeCommentEntity;
import dev.ragnarok.fenrir.db.model.entity.feedback.LikeEntity;
import dev.ragnarok.fenrir.db.model.entity.feedback.MentionCommentEntity;
import dev.ragnarok.fenrir.db.model.entity.feedback.MentionEntity;
import dev.ragnarok.fenrir.db.model.entity.feedback.NewCommentEntity;
import dev.ragnarok.fenrir.db.model.entity.feedback.PostFeedbackEntity;
import dev.ragnarok.fenrir.db.model.entity.feedback.ReplyCommentEntity;
import dev.ragnarok.fenrir.db.model.entity.feedback.UsersEntity;
import dev.ragnarok.fenrir.domain.IFeedbackInteractor;
import dev.ragnarok.fenrir.domain.IOwnersRepository;
import dev.ragnarok.fenrir.domain.mappers.Dto2Entity;
import dev.ragnarok.fenrir.domain.mappers.Dto2Model;
import dev.ragnarok.fenrir.domain.mappers.Entity2Model;
import dev.ragnarok.fenrir.domain.mappers.FeedbackEntity2Model;
import dev.ragnarok.fenrir.model.AnswerVKOfficialList;
import dev.ragnarok.fenrir.model.Owner;
import dev.ragnarok.fenrir.model.criteria.NotificationsCriteria;
import dev.ragnarok.fenrir.model.feedback.Feedback;
import dev.ragnarok.fenrir.util.Pair;
import dev.ragnarok.fenrir.util.Utils;
import dev.ragnarok.fenrir.util.VKOwnIds;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Single;

public class FeedbackInteractor implements IFeedbackInteractor {

    private final IStorages cache;
    private final INetworker networker;
    private final IOwnersRepository ownersRepository;

    public FeedbackInteractor(IStorages cache, INetworker networker, IOwnersRepository ownersRepository) {
        this.cache = cache;
        this.networker = networker;
        this.ownersRepository = ownersRepository;
    }

    private static void populateOwnerIds(VKOwnIds ids, FeedbackEntity dbo) {
        fillCommentOwnerIds(ids, dbo.getReply());

        if (dbo instanceof CopyEntity) {
            populateOwnerIds(ids, (CopyEntity) dbo);
        } else if (dbo instanceof LikeCommentEntity) {
            populateOwnerIds(ids, (LikeCommentEntity) dbo);
        } else if (dbo instanceof LikeEntity) {
            populateOwnerIds(ids, (LikeEntity) dbo);
        } else if (dbo instanceof MentionCommentEntity) {
            populateOwnerIds(ids, (MentionCommentEntity) dbo);
        } else if (dbo instanceof MentionEntity) {
            populateOwnerIds(ids, (MentionEntity) dbo);
        } else if (dbo instanceof NewCommentEntity) {
            populateOwnerIds(ids, (NewCommentEntity) dbo);
        } else if (dbo instanceof PostFeedbackEntity) {
            populateOwnerIds(ids, (PostFeedbackEntity) dbo);
        } else if (dbo instanceof ReplyCommentEntity) {
            populateOwnerIds(ids, (ReplyCommentEntity) dbo);
        } else if (dbo instanceof UsersEntity) {
            populateOwnerIds(ids, (UsersEntity) dbo);
        }
    }

    private static void populateOwnerIds(VKOwnIds ids, UsersEntity dbo) {
        ids.appendAll(dbo.getOwners());
    }

    private static void populateOwnerIds(VKOwnIds ids, ReplyCommentEntity dbo) {
        Entity2Model.fillOwnerIds(ids, dbo.getCommented());
        Entity2Model.fillOwnerIds(ids, dbo.getFeedbackComment());
        Entity2Model.fillOwnerIds(ids, dbo.getOwnComment());
    }

    private static void populateOwnerIds(VKOwnIds ids, PostFeedbackEntity dbo) {
        Entity2Model.fillOwnerIds(ids, dbo.getPost());
    }

    private static void populateOwnerIds(VKOwnIds ids, NewCommentEntity dbo) {
        Entity2Model.fillOwnerIds(ids, dbo.getComment());
        Entity2Model.fillOwnerIds(ids, dbo.getCommented());
    }

    private static void populateOwnerIds(VKOwnIds ids, MentionEntity dbo) {
        Entity2Model.fillOwnerIds(ids, dbo.getWhere());
    }

    private static void populateOwnerIds(VKOwnIds ids, MentionCommentEntity dbo) {
        Entity2Model.fillOwnerIds(ids, dbo.getCommented());
        Entity2Model.fillOwnerIds(ids, dbo.getWhere());
    }

    private static void populateOwnerIds(VKOwnIds ids, LikeEntity dbo) {
        Entity2Model.fillOwnerIds(ids, dbo.getLiked());
        ids.appendAll(dbo.getLikesOwnerIds());
    }

    private static void populateOwnerIds(VKOwnIds ids, LikeCommentEntity dbo) {
        Entity2Model.fillOwnerIds(ids, dbo.getLiked());
        Entity2Model.fillOwnerIds(ids, dbo.getCommented());
        ids.appendAll(dbo.getLikesOwnerIds());
    }

    private static void populateOwnerIds(VKOwnIds ids, CopyEntity dbo) {
        for (IdPairEntity i : dbo.getCopies().getPairDbos()) {
            ids.append(i.getOwnerId());
        }

        Entity2Model.fillOwnerIds(ids, dbo.getCopied());
    }

    @Override
    public Single<List<Feedback>> getCachedFeedbacks(int accountId) {
        NotificationsCriteria criteria = new NotificationsCriteria(accountId);
        return getCachedFeedbacksByCriteria(criteria);
    }

    @Override
    public Single<AnswerVKOfficialList> getOfficial(int accountId, Integer count, Integer startFrom) {
        return networker.vkDefault(accountId)
                .notifications()
                .getOfficial(count, startFrom, null, null, null)
                .map(response -> response);
    }

    @Override
    public Single<Pair<List<Feedback>, String>> getActualFeedbacks(int accountId, int count, String startFrom) {
        return networker.vkDefault(accountId)
                .notifications()
                .get(count, startFrom, null, null, null)
                .flatMap(response -> {
                    List<VkApiBaseFeedback> dtos = Utils.listEmptyIfNull(response.notifications);
                    List<FeedbackEntity> dbos = new ArrayList<>(dtos.size());

                    VKOwnIds ownIds = new VKOwnIds();

                    for (VkApiBaseFeedback dto : dtos) {
                        FeedbackEntity dbo = Dto2Entity.buildFeedbackDbo(dto);
                        populateOwnerIds(ownIds, dbo);
                        dbos.add(dbo);
                    }

                    OwnerEntities ownerEntities = Dto2Entity.mapOwners(response.profiles, response.groups);
                    List<Owner> owners = Dto2Model.transformOwners(response.profiles, response.groups);

                    return cache.notifications()
                            .insert(accountId, dbos, ownerEntities, isEmpty(startFrom))
                            .flatMap(ints -> ownersRepository
                                    .findBaseOwnersDataAsBundle(accountId, ownIds.getAll(), IOwnersRepository.MODE_ANY, owners)
                                    .map(ownersBundle -> {
                                        List<Feedback> feedbacks = new ArrayList<>(dbos.size());

                                        for (FeedbackEntity dbo : dbos) {
                                            feedbacks.add(FeedbackEntity2Model.buildFeedback(dbo, ownersBundle));
                                        }

                                        return Pair.Companion.create(feedbacks, response.nextFrom);
                                    }));
                });
    }

    @Override
    public Completable maskAaViewed(int accountId) {
        return networker.vkDefault(accountId)
                .notifications()
                .markAsViewed()
                .ignoreElement();
    }

    private Single<List<Feedback>> getCachedFeedbacksByCriteria(NotificationsCriteria criteria) {
        return cache.notifications()
                .findByCriteria(criteria)
                .flatMap(dbos -> {
                    VKOwnIds ownIds = new VKOwnIds();

                    for (FeedbackEntity dbo : dbos) {
                        populateOwnerIds(ownIds, dbo);
                    }

                    return ownersRepository.findBaseOwnersDataAsBundle(criteria.getAccountId(), ownIds.getAll(), IOwnersRepository.MODE_ANY)
                            .map(owners -> {
                                List<Feedback> feedbacks = new ArrayList<>(dbos.size());
                                for (FeedbackEntity dbo : dbos) {
                                    feedbacks.add(FeedbackEntity2Model.buildFeedback(dbo, owners));
                                }
                                return feedbacks;
                            });
                });
    }
}

File: app/src/main/java/dev/ragnarok/fenrir/domain/impl/FeedbackInteractor.kt
package dev.ragnarok.fenrir.domain.impl

import dev.ragnarok.fenrir.api.interfaces.INetworker
import dev.ragnarok.fenrir.db.interfaces.IStorages
import dev.ragnarok.fenrir.db.model.entity.feedback.*
import dev.ragnarok.fenrir.domain.IFeedbackInteractor
import dev.ragnarok.fenrir.domain.IOwnersRepository
import dev.ragnarok.fenrir.domain.mappers.Dto2Entity.buildFeedbackDbo
import dev.ragnarok.fenrir.domain.mappers.Dto2Entity.mapOwners
import dev.ragnarok.fenrir.domain.mappers.Dto2Model.transformOwners
import dev.ragnarok.fenrir.domain.mappers.Entity2Model.fillCommentOwnerIds
import dev.ragnarok.fenrir.domain.mappers.Entity2Model.fillOwnerIds
import dev.ragnarok.fenrir.domain.mappers.FeedbackEntity2Model.buildFeedback
import dev.ragnarok.fenrir.model.AnswerVKOfficialList
import dev.ragnarok.fenrir.model.criteria.NotificationsCriteria
import dev.ragnarok.fenrir.model.feedback.Feedback
import dev.ragnarok.fenrir.util.Pair
import dev.ragnarok.fenrir.util.Pair.Companion.create
import dev.ragnarok.fenrir.util.Utils.listEmptyIfNull
import dev.ragnarok.fenrir.util.VKOwnIds
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Single

class FeedbackInteractor(
    private val cache: IStorages,
    private val networker: INetworker,
    private val ownersRepository: IOwnersRepository
) : IFeedbackInteractor {
    override fun getCachedFeedbacks(accountId: Int): Single<List<Feedback>> {
        val criteria = NotificationsCriteria(accountId)
        return getCachedFeedbacksByCriteria(criteria)
    }

    override fun getOfficial(
        accountId: Int,
        count: Int?,
        startFrom: Int?
    ): Single<AnswerVKOfficialList> {
        return networker.vkDefault(accountId)
            .notifications()
            .getOfficial(count, startFrom, null, null, null)
            .map { response -> response }
    }

    override fun hide(accountId: Int, query: String?): Completable {
        return networker.vkDefault(accountId)
            .notifications()
            .hide(query)
            .ignoreElement()
    }

    override fun getActualFeedbacks(
        accountId: Int,
        count: Int,
        startFrom: String?
    ): Single<Pair<List<Feedback>, String>> {
        return networker.vkDefault(accountId)
            .notifications()[count, startFrom, null, null, null]
            .flatMap { response ->
                val dtos = listEmptyIfNull(response.notifications)
                val dbos: MutableList<FeedbackEntity> = ArrayList(dtos.size)
                val ownIds = VKOwnIds()
                for (dto in dtos) {
                    val dbo = buildFeedbackDbo(dto)
                    populateOwnerIds(ownIds, dbo)
                    dbos.add(dbo)
                }
                val ownerEntities = mapOwners(response.profiles, response.groups)
                val owners = transformOwners(response.profiles, response.groups)
                cache.notifications()
                    .insert(accountId, dbos, ownerEntities, startFrom.isNullOrEmpty())
                    .flatMap {
                        ownersRepository
                            .findBaseOwnersDataAsBundle(
                                accountId,
                                ownIds.all,
                                IOwnersRepository.MODE_ANY,
                                owners
                            )
                            .map {
                                val feedbacks: MutableList<Feedback> = ArrayList(dbos.size)
                                for (dbo in dbos) {
                                    feedbacks.add(buildFeedback(dbo, it))
                                }
                                create(feedbacks, response.nextFrom)
                            }
                    }
            }
    }

    override fun maskAaViewed(accountId: Int): Completable {
        return networker.vkDefault(accountId)
            .notifications()
            .markAsViewed()
            .ignoreElement()
    }

    private fun getCachedFeedbacksByCriteria(criteria: NotificationsCriteria): Single<List<Feedback>> {
        return cache.notifications()
            .findByCriteria(criteria)
            .flatMap { dbos ->
                val ownIds = VKOwnIds()
                for (dbo in dbos) {
                    populateOwnerIds(ownIds, dbo)
                }
                ownersRepository.findBaseOwnersDataAsBundle(
                    criteria.accountId,
                    ownIds.all,
                    IOwnersRepository.MODE_ANY
                )
                    .map<List<Feedback>> {
                        val feedbacks: MutableList<Feedback> = ArrayList(dbos.size)
                        for (dbo in dbos) {
                            feedbacks.add(buildFeedback(dbo, it))
                        }
                        feedbacks
                    }
            }
    }

    companion object {
        private fun populateOwnerIds(ids: VKOwnIds, dbo: FeedbackEntity) {
            fillCommentOwnerIds(ids, dbo.reply)
            when (dbo) {
                is CopyEntity -> {
                    populateOwnerIds(ids, dbo)
                }
                is LikeCommentEntity -> {
                    populateOwnerIds(ids, dbo)
                }
                is LikeEntity -> {
                    populateOwnerIds(ids, dbo)
                }
                is MentionCommentEntity -> {
                    populateOwnerIds(ids, dbo)
                }
                is MentionEntity -> {
                    populateOwnerIds(ids, dbo)
                }
                is NewCommentEntity -> {
                    populateOwnerIds(ids, dbo)
                }
                is PostFeedbackEntity -> {
                    populateOwnerIds(ids, dbo)
                }
                is ReplyCommentEntity -> {
                    populateOwnerIds(ids, dbo)
                }
                is UsersEntity -> {
                    populateOwnerIds(ids, dbo)
                }
            }
        }

        private fun populateOwnerIds(ids: VKOwnIds, dbo: UsersEntity) {
            ids.appendAll(dbo.owners)
        }

        private fun populateOwnerIds(ids: VKOwnIds, dbo: ReplyCommentEntity) {
            fillOwnerIds(ids, dbo.commented)
            fillOwnerIds(ids, dbo.feedbackComment)
            fillOwnerIds(ids, dbo.ownComment)
        }

        private fun populateOwnerIds(ids: VKOwnIds, dbo: PostFeedbackEntity) {
            fillOwnerIds(ids, dbo.post)
        }

        private fun populateOwnerIds(ids: VKOwnIds, dbo: NewCommentEntity) {
            fillOwnerIds(ids, dbo.comment)
            fillOwnerIds(ids, dbo.commented)
        }

        private fun populateOwnerIds(ids: VKOwnIds, dbo: MentionEntity) {
            fillOwnerIds(ids, dbo.where)
        }

        private fun populateOwnerIds(ids: VKOwnIds, dbo: MentionCommentEntity) {
            fillOwnerIds(ids, dbo.commented)
            fillOwnerIds(ids, dbo.where)
        }

        private fun populateOwnerIds(ids: VKOwnIds, dbo: LikeEntity) {
            fillOwnerIds(ids, dbo.liked)
            ids.appendAll(dbo.likesOwnerIds)
        }

        private fun populateOwnerIds(ids: VKOwnIds, dbo: LikeCommentEntity) {
            fillOwnerIds(ids, dbo.liked)
            fillOwnerIds(ids, dbo.commented)
            ids.appendAll(dbo.likesOwnerIds)
        }

        private fun populateOwnerIds(ids: VKOwnIds, dbo: CopyEntity) {
            for (i in dbo.copies.pairDbos) {
                ids.append(i.ownerId)
            }
            fillOwnerIds(ids, dbo.copied)
        }
    }
}